src/NumericalAlgorithms/Spectral/SwshCoefficients.hpp

```diff
@@ -125,59 +133,183 @@ struct DestroySharpAlm {
+
+    LibsharpCoefficientInfo operator*() const noexcept {
+      // permit dereferencing the iterator only if the return represents a
+      // viable location in the coefficient vector
+      ASSERT(l_ <= l_max_ && m_ <= l_max_, "coefficients iterator overflow");
+      size_t offset = ((3 + 2 * l_max_ - m_) * m_) / 2 + l_ - m_;
+      return LibsharpCoefficientInfo{
+          offset,
+          offset +
+              Spectral::Swsh::size_of_libsharp_coefficient_vector(l_max_) / 2,
+          l_max_, l_, m_};
+    }
+    /// advance the iterator by one position (prefix)
+    CoefficientsConstIterator& operator++() noexcept {
+      // permit altering the iterator only if the new value is between the
+      // anticipated begin and end, inclusive
+      ASSERT(l_ <= l_max_ && m_ <= l_max_, "coefficients iterator overflow");
+      if (l_ == l_max_) {
+        ++m_;
+        l_ = m_;
+      } else {
+        ++l_;
+      }
+      return *this;
+    };
+
+    /// retreat the iterator by one position (prefix)
+    CoefficientsConstIterator& operator--() noexcept {
+      // permit altering the iterator only if the new value is between the
+      // anticipated begin and end, inclusive
+      ASSERT(l_ <= l_max_ + 1 && m_ <= l_max_ + 1,
+             "coefficients iterator overflow");
+      if (l_ == m_) {
+        --m_;
+        l_ = l_max_;
+      } else {
+        --l_;
+      }
+      return *this;
+    }
+
+    // @{
+    /// (In)Equivalence checks the object as well as the l and m current
+    /// position.
+    bool operator==(const CoefficientsConstIterator& rhs) const noexcept {
+      return m_ == rhs.m_ and l_ == rhs.l_ and l_max_ == rhs.l_max_;
+    }
+    bool operator!=(const CoefficientsConstIterator& rhs) const noexcept {
+      return not(*this == rhs);
+    }
+    // @}
+
+   private:
+    size_t m_;
+    size_t l_;
+    size_t l_max_;
+  };
+
+  explicit CoefficientsMetadata(size_t l_max) noexcept;
+
+  ~CoefficientsMetadata() = default;
+  CoefficientsMetadata() = default;
+  CoefficientsMetadata(const CoefficientsMetadata&) = delete;
+  CoefficientsMetadata(CoefficientsMetadata&&) = default;
+  CoefficientsMetadata& operator=(const CoefficientsMetadata&) = delete;
+  CoefficientsMetadata& operator=(CoefficientsMetadata&&) = default;
   sharp_alm_info* get_sharp_alm_info() const noexcept {
     return alm_info_.get();
   }
 
   size_t l_max() const noexcept { return l_max_; }
 
+  /// returns the number of (complex) entries in a libsharp-compatible
+  /// coefficients vector. This includes the factor of 2 associated with needing
+  /// to store both the transform of the real and imaginary parts, so is the
+  /// full size of the result of a libsharp swsh transform.
+  size_t size() const noexcept {
+    return size_of_libsharp_coefficient_vector(l_max_);
+  }
+
+  // @{
+  /// \brief Get a bidirectional iterator to the start of the series of modes.
+  CoefficientsMetadata::CoefficientsConstIterator begin() const noexcept {
```

Comment:
To make sure I understand: these iterators leave the underlying data immutable, right? So I could not change the value of a particular coefficient using these iterators.

Reply:
Note that these iterators dereference to objects which just contain l, m, and offests into a vector. So, the mutability of the underlying data is determined by the vector you use the offset to index.

Reply:
Hmm, okay so this is more like a `CoefficientsIndexIterator` then. I'd prefer a name change that makes it clear you are not going to get a modal coefficient when dereferencing the iterator. What do you think?

Reply:
Yeah, I'm happy making it `CoefficientsIndexIterator`